sql/src/main/java/io/crate/execution/IncrementalPageBucketReceiver.java

```diff
@@ -0,0 +1,120 @@
+import io.crate.data.RowConsumer;
+import io.crate.data.SentinelRow;
+import io.crate.execution.jobs.PageBucketReceiver;
+import io.crate.execution.jobs.PageResultListener;
+
+import javax.annotation.Nonnull;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.stream.Collector;
+
+public class IncrementalPageBucketReceiver<T> implements PageBucketReceiver {
+
+    private final Function<T, Iterable<Row>> finisher;
+    private final BiConsumer<T, Row> accumulator;
+    private final T state;
+    private final AtomicInteger remainingUpstreams;
+    private final RowConsumer consumer;
+    private final CompletableFuture<?> processingFuture = new CompletableFuture<>();
+    private final Streamer<?>[] streamers;
+
+    private volatile Throwable lastThrowable = null;
+
+    public IncrementalPageBucketReceiver(Collector<Row, T, Iterable<Row>> collector,
+                                         RowConsumer rowConsumer,
+                                         Streamer<?>[] streamers,
+                                         int upstreamsCount) {
+        this.state = collector.supplier().get();
+        this.accumulator = collector.accumulator();
+        this.finisher = collector.finisher();
+        this.consumer = rowConsumer;
+        this.streamers = streamers;
+        this.remainingUpstreams = new AtomicInteger(upstreamsCount);
+    }
+
+    @Override
+    public void setBucket(int bucketIdx, Bucket rows, boolean isLast, PageResultListener pageResultListener) {
+        synchronized (state) {
+            try {
+                for (Row row : rows) {
+                    accumulator.accept(state, row);
+                }
+            } catch (Throwable e) {
+                lastThrowable = e;
+            }
+        }
+
+        pageResultListener.needMore(!isLast);
+        if (isLast) {
+            if (remainingUpstreams.decrementAndGet() == 0) {
+                consumeRows();
+                if (lastThrowable == null) {
+                    processingFuture.complete(null);
+                } else {
+                    processingFuture.completeExceptionally(lastThrowable);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void releasePageResultListeners() {
+    }
+
+    @Override
+    public void killed(int bucketIdx, Throwable throwable) {
+        kill(throwable);
+    }
+
+    @Override
+    public Streamer<?>[] streamers() {
+        return streamers;
+    }
+
+    @Override
+    public CompletableFuture<?> completionFuture() {
+        return processingFuture;
+    }
+
+    @Override
+    public void consumeRows() {
+        consumer.accept(InMemoryBatchIterator.of(finisher.apply(state), SentinelRow.SENTINEL), lastThrowable);
+    }
+
+    @Override
+    public void kill(@Nonnull Throwable t) {
+        lastThrowable = t;
+        consumer.accept(null, t);
```

Comment:
Is it really safe to call `consumer.accept` here? Afaik the contract is that `accept` must only be called once. If `consumeRows` has been called before `kill` is invoked it might have been called already.

One way to avoid this might be to call the `consumer.accept` early in the constructor with a `CollectingBatchIterator` that has a `Supplier` which simply returns the `processingFuture` (and the future would then contain the `Iterable<Row>` result)